dev/tools/merge-pr.sh

```diff
@@ -1,50 +1,126 @@
 
-#TODO: check arguments and show usage if relevant
+check_util() {
+if ! which $1 > /dev/null; then
+  echo error: this script requires the $1 command line utility
+  exit 1
+fi
+}
 
-PR=$1
+ask_confirmation() {
+  read -p "Continue anyway? [y/N] " -n 1 -r
+  echo
+  if [[ ! $REPLY =~ ^[Yy]$ ]]
+  then
+    exit 1
+  fi
+}
 
-CURRENT_LOCAL_BRANCH=$(git rev-parse --abbrev-ref HEAD)
-REMOTE=$(git config --get "branch.$CURRENT_LOCAL_BRANCH.remote")
-git fetch "$REMOTE" "refs/pull/$PR/head"
+RED="\033[31m"
+BLACK="\033[30m"
+GREEN="\033[32m"
+BLUE="\033[34m"
+YELLOW="\033[33m"
+info() {
+  echo -e "${GREEN}info:${BLACK}" $@ ${BLACK}
+}
+error() {
+  echo -e "${RED}error:${BLACK}" $@ ${BLACK}
+}
+warning() {
+  echo -e "${YELLOW}warning:${BLACK}" $@ ${BLACK}
+}
 
-API=https://api.github.com/repos/coq/coq
+check_util jq
+check_util curl
+check_util git
+check_util gpg
+
+# command line parsing
+
+if [ $# != 1 ]; then
+  echo usage: $0 PR-number
+  exit 1
+fi
+
+if [[ "$1" =~ ^[0-9][0-9]*$ ]]; then
+  PR=$1
+else
+  error $1 is not a number
+  exit 1
+fi
+
+# Fetching PR metadata
+
+TITLE=$(curl -s "$API/pulls/$PR" | jq -r '.title')
+info title for PR $PR is ${BLUE}$TITLE
 
 BASE_BRANCH=$(curl -s "$API/pulls/$PR" | jq -r '.base.label')
+info PR $PR targets branch ${BLUE}$BASE_BRANCH
+
+CURRENT_LOCAL_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+info you are merging in ${BLUE}$CURRENT_LOCAL_BRANCH
+
+REMOTE=$(git config --get "branch.$CURRENT_LOCAL_BRANCH.remote")
+info remote for $CURRENT_LOCAL_BRANCH is ${BLUE}$REMOTE
 
+info fetching from $REMOTE the PR
+git remote update "$REMOTE"
+if ! git ls-remote "$REMOTE" | grep pull >/dev/null; then
+  error remote $REMOTE is not configured to fetch pull requests
+  error run: git config remote.$REMOTE.fetch +refs/pull/*/head:refs/remotes/$REMOTE/pr/*
+  exit 1
+fi
+git fetch "$REMOTE" "refs/pull/$PR/head"
 COMMIT=$(git rev-parse FETCH_HEAD)
-STATUS=$(curl -s "$API/commits/$COMMIT/status" | jq -r '.state')
+info commit for for PR $PR is ${BLUE}$COMMIT
+
+# Sanity check: merge to a different branch
 
 if [ "$BASE_BRANCH" != "coq:$CURRENT_LOCAL_BRANCH" ]; then
-  echo "Wrong base branch"
-  read -p "Bypass? [y/N] " -n 1 -r
-  echo
-  if [[ ! $REPLY =~ ^[Yy]$ ]]
-  then
-    exit 1
-  fi
+  echo "error: PR requests merge in ${BLUE}$BASE_BRANCH${BLACK} but you are merging in ${BLUE}$CURRENT_LOCAL_BRANCH"
+  ask_confirmation
 fi;
 
+# Sanity check: CI failed
+
+STATUS=$(curl -s "$API/commits/$COMMIT/status" | jq -r '.state')
+
 if [ "$STATUS" != "success" ]; then
-  echo "CI status is \"$STATUS\""
-  read -p "Bypass? [y/N] " -n 1 -r
-  echo
-  if [[ ! $REPLY =~ ^[Yy]$ ]]
-  then
-    exit 1
-  fi
+  error CI failure on ${BLUE}$(curl -s "$API/commits/$COMMIT/status" |
+    jq -r -c '.statuses|map(select(.state == "failure"))|map(.context)')
+  ask_confirmation
 fi;
 
-git merge -S --no-ff FETCH_HEAD -m "Merge PR #$PR: $(curl -s "$API/pulls/$PR" | jq -r '.title')" -e
+# Sanity check: has labels named "needs:"
```

Comment:
Cool! Could you do the same for milestones? I'm tired of having to set milestones after a PR has been merged.

Reply:
Oh and also check that the PR has at least one "kind: " label :)

Reply:
>Oh and also check that the PR has at least one "kind: " label :)

Is that really something we want to check?
I don't understand what you want to happen with milestones as well.

Reply:
What I've been enforcing manually since the 8.7 release (or even since the 8.6.1 release) is that all merged PRs are in a milestone and have a kind. This means that one can browse all the changes that happened between two versions by checking the corresponding milestone (and the kind labels show what kind of change a PR represents). This is like a mega CHANGES file but with much more information.

Reply:
A further improvement that would be really great would be for the script to allow setting the milestone directly (without going back in the web interface). This should be doable but only with write access, which would require people to first get a personal API token, etc. Much more complex thus!

Reply:
I'll add a sanity check for kind: too